state.go

```diff
@@ -0,0 +1,308 @@
+// Copyright 2016 Mender Software AS
+//
+//    Licensed under the Apache License, Version 2.0 (the "License");
+//    you may not use this file except in compliance with the License.
+//    You may obtain a copy of the License at
+//
+//        http://www.apache.org/licenses/LICENSE-2.0
+//
+//    Unless required by applicable law or agreed to in writing, software
+//    distributed under the License is distributed on an "AS IS" BASIS,
+//    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//    See the License for the specific language governing permissions and
+//    limitations under the License.
+package main
+
+import (
+	"io"
+	"time"
+
+	"github.com/mendersoftware/log"
+	"github.com/pkg/errors"
+)
+
+type State interface {
+	// Perform state action, returns next state and boolean flag indicating if
+	// execution was cancelled or not
+	Handle(c Controler) (State, bool)
```

Comment:
It's not really changing the state, it's only returning a new state. The state implements a strategy that works with the controller to do stuff. In fact it's a know design pattern - aka. state design pattern. Although, I did not implement a separate controller method that executes the state action (instead we just `GetState().Handle(..)` in daemon).